woocommerce/woocommerce-ios
# [Mobile Payments] Add support for querying the WCPay account endpoint

**Ready for Review - Don't let the size scare you. Roughly half of it is JSON responses for the tests.**

⚠️ This PR is against the feature branch implementing support for Card Present Payments, not against develop ⚠️

Partially addresses https://github.com/woocommerce/woocommerce-ios/issues/3839

## Changes

Adds a remote for fetching selected WCPay account details from the self-hosted site. This is needed in order to know
- if we should allow the merchant to scan for readers or not (if they don't have an account, token fetching will fail)
- what statement descriptor to use in payment intents (see https://github.com/woocommerce/woocommerce-ios/pull/3879 )
- if the merchant has an account, but it has been rejected (i.e. such that we should not attempt card present payments)
- if the merchant has an account, but it has been restricted (i.e. such that we should not attempt card present payments - and we could display a prompt encouraging them to rectify that)
- what currencies the site supports
- what country the underlying stripe account is in

Note: Restricts query to _fields we need
Note: Requires https://github.com/Automattic/woocommerce-payments/pull/1403 to actually retrieve account

## Testing Instructions

- Ensure WCPayRemoteTests (now 13) all pass

## Advanced Testing Instructions
- Since this is the networking layer, there isn't a way to trigger this from the UI (yet), but if you really want to...
- Install a WooCommerce Payments plugin using this branch https://github.com/woocommerce/woocommerce-ios/issues/3839
- Try any of the following: A site with WCPay active but no on-boarding completed, a site with WCPay account using a dev mode/test account only, or a site with WCPay active using a live for-reals $$$$ account
- Hack something like SettingStore.swift to set a breakpoint to see the response for your site and the mapping to WCPayAccount, e.g.

```diff
diff --git a/Yosemite/Yosemite/Stores/SettingStore.swift b/Yosemite/Yosemite/Stores/SettingStore.swift
index cb5ac8b18..4c2cdd811 100644
--- a/Yosemite/Yosemite/Stores/SettingStore.swift
+++ b/Yosemite/Yosemite/Stores/SettingStore.swift
@@ -8,6 +8,7 @@ import Storage
 public class SettingStore: Store {
     private let siteSettingsRemote: SiteSettingsRemote
     private let siteAPIRemote: SiteAPIRemote
+    private let wcPayRemote: WCPayRemote
 
     private lazy var sharedDerivedStorage: StorageType = {
         return storageManager.newDerivedStorage()
@@ -16,6 +17,7 @@ public class SettingStore: Store {
     public override init(dispatcher: Dispatcher, storageManager: StorageManagerType, network: Network) {
         self.siteSettingsRemote = SiteSettingsRemote(network: network)
         self.siteAPIRemote = SiteAPIRemote(network: network)
+        self.wcPayRemote = WCPayRemote(network:network)
         super.init(dispatcher: dispatcher, storageManager: storageManager, network: network)
     }
 
@@ -62,6 +64,9 @@ private extension SettingStore {
                 onCompletion(nil)
             }
         }
+        wcPayRemote.loadAccount(for: siteID) { [weak self] (result) in
+
+        }
     }
 
     /// Synchronizes the product site settings associated with the provided Site ID (if any!).
```

Update release notes:
